src/Microsoft.AspNetCore.Mvc.Core/Internal/MvcEndpointDataSource.cs

```diff
@@ -298,20 +299,18 @@ public Task RouteAsync(RouteContext context)
                 return invoker.InvokeAsync();
             };
 
+            var defaults = new RouteValueDictionary(nonInlineDefaults);
+            EnsureRequiredValuesInDefaults(action.RouteValues, defaults);
+
             var metadataCollection = BuildEndpointMetadata(action, routeName, source);
             var endpoint = new MatcherEndpoint(
                 next => invokerDelegate,
-                template,
-                new RouteValueDictionary(nonInlineDefaults),
+                RoutePatternFactory.Parse(template, defaults, constraints: null),
```

Comment:
Wouldn't this throw if defaults were specified inline as well as non-inline?

Reply:
Like would this hit https://github.com/aspnet/Routing/blob/release/2.2/src/Microsoft.AspNetCore.Routing/Patterns/RoutePatternFactory.cs#L215?

Reply:
They can't be duplicates of each other - the templates we pass in here never have the 'action' and 'controller' as parameters. This was actually a problem in the benchmark @JamesNK - what it was testing for attribute routing is something that will (and should) throw.